jms-core/src/main/java/io/micronaut/jms/listener/JMSListener.java

```diff
@@ -0,0 +1,140 @@
+
+    /***
+     * Creates a {@link JMSListener} instance. This instance will not begin listening for messages until
+     *  {@link JMSListener#start()} is called. The provided {@param session}'s parent {@link javax.jms.Connection}
+     *  must be started ({@link javax.jms.Connection#start()}) for the message listener to receive messages.
+     *
+     * @param session - the {@link Session} for the messages to be consumed on
+     * @param delegate - the listener logic to be invoked. All concurrency, success, and error handling is provided.
+     *                 This {@link MessageListener} should extract the necessary data from the {@link javax.jms.Message}
+     *                 and perform application specific logic.
+     * @param destinationType - the {@link JMSDestinationType} of the target destination
+     * @param destination - the name of the target destination
+     * @param executor - the {@link ExecutorService} to perform the message handling logic on. The message handling, including
+     *                 success and error handling, is performed on threads managed by this {@param executor}.
+     */
+    public JMSListener(Session session, MessageListener delegate, JMSDestinationType destinationType, String destination, ExecutorService executor) {
+        this.session = session;
+        this.delegate = delegate;
+        this.destinationType = destinationType;
+        this.destination = destination;
+        this.executor = executor;
+    }
+
+    public void addSuccessHandlers(JMSListenerSuccessHandler... handlers) {
+        this.addSuccessHandlers(Arrays.asList(handlers));
+    }
+
+    public void addSuccessHandlers(Collection<JMSListenerSuccessHandler> handlers) {
+        successHandlers.addAll(handlers);
+    }
+
+    public void addErrorHandlers(JMSListenerErrorHandler... handlers) {
+        this.addErrorHandlers(Arrays.asList(handlers));
+    }
+
+    public void addErrorHandlers(Collection<JMSListenerErrorHandler> handlers) {
+        errorHandlers.addAll(handlers);
+    }
+
+    public void start() throws JMSException {
+        final MessageConsumer consumer = session.createConsumer(lookupDestination(destinationType, destination, session));
+        consumer.setMessageListener((msg) -> executor.submit(() -> {
+            try {
+                delegate.onMessage(msg);
+                Throwable ex = new Throwable();
+                successHandlers.forEach(handler -> {
+                    try {
+                        handler.handle(session, msg);
+                    } catch (JMSException e) {
+                        LOGGER.error("Failed to handle successful message receive", e);
+                        ex.addSuppressed(e);
+                    }
+                });
+                if (ex.getSuppressed().length > 0) {
+                    throw ex;
+                }
+            } catch (Throwable e) {
+                errorHandlers.forEach(handler -> handler.handle(session, msg, e));
```

Comment:
should just move this to line 117 in place of `throw ex`, there's a non-zero cost to throwing exceptions (filling in the stack trace, etc.)